clippy_lints/src/excessive_nesting.rs

```diff
@@ -0,0 +1,329 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use rustc_ast::{
+    node_id::NodeId,
+    ptr::P,
+    visit::{FnKind, Visitor},
+    Arm, AssocItemKind, Block, Expr, ExprKind, Inline, Item, ItemKind, Local, LocalKind, ModKind, ModSpans, Pat,
+    PatKind, Stmt, StmtKind,
+};
+use rustc_lint::{EarlyContext, EarlyLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::Span;
+use thin_vec::ThinVec;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Checks for blocks which are indented beyond a certain threshold.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// It can severely hinder readability. The default is very generous; if you
+    /// exceed this, it's a sign you should refactor.
+    ///
+    /// ### Known issues
+    ///
+    /// Nested inline modules will all be linted, rather than just the outermost one
+    /// that applies. This makes the output a bit verbose.
+    ///
+    /// ### Example
+    /// An example clippy.toml configuration:
+    /// ```toml
+    /// # clippy.toml
+    /// excessive-nesting-threshold = 3
+    /// ```
+    /// lib.rs:
+    /// ```rust,ignore
+    /// pub mod a {
+    ///     pub struct X;
+    ///     impl X {
+    ///         pub fn run(&self) {
+    ///             if true {
+    ///                 // etc...
+    ///             }
+    ///         }
+    ///     }
+    /// }
+    /// Use instead:
+    /// a.rs:
+    /// ```rust,ignore
+    /// fn private_run(x: &X) {
+    ///     if true {
+    ///         // etc...
+    ///     }
+    /// }
+    ///
+    /// pub struct X;
+    /// impl X {
+    ///     pub fn run(&self) {
+    ///         private_run(self);
+    ///     }
+    /// }
+    /// ```
+    #[clippy::version = "1.70.0"]
+    pub EXCESSIVE_NESTING,
+    restriction,
+    "checks for blocks nested beyond a certain threshold"
+}
+impl_lint_pass!(ExcessiveNesting => [EXCESSIVE_NESTING]);
+
+#[derive(Clone, Copy)]
+pub struct ExcessiveNesting {
+    pub excessive_nesting_threshold: u64,
+}
+
+impl EarlyLintPass for ExcessiveNesting {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, item: &Item) {
+        let conf = self;
+        NestingVisitor {
+            conf,
+            cx,
+            nest_level: 0,
+        }
+        .visit_item(item);
+    }
```

Comment:
I noticed this while testing, hence the note under "Known issues" (though I thought it was only specific to modules). At first, I was thinking of using `check_crate` instead but thought it wasn't necessary since all it was doing was calling `visit_item` for each item anyway, but I'll change it to that (plus `Inline::No`) once I have the chance. Thanks

Reply:
Fixed, also removed the known issues section since it doesn't apply anymore